### PR TITLE
[SPARK-29997][WEBUI][FOLLOWUP] Refactor code for job description of empty jobs

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

Refactor the code brought by #26637 .
No more dummy StageInfo and its side-effects are needed at all.
This change also enable users to set job description to empty jobs though.

### Why are the changes needed?

The previous approach introduced dummy StageInfo and this causes side-effects.

### Does this PR introduce any user-facing change?

Yes. Description set by user will be shown in the AllJobsPage.

### How was this patch tested?

Manual test and newly added unit test.

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
This looks like a revert of the previous PR mostly.

**Comment by dongjoon-hyun** [comment]
Shall we do this as a follow up of SPARK-29997 ? Or, you can revert SPARK-29997 and do this cleanly.

**Comment by sarutak** [comment]
> This looks like a revert of the previous PR mostly.
At first, I've tried to fix the issue which the title of this PR says and I noticed that I can use `setDescription` to refactor the previous PR.

> Shall we do this as a follow up of SPARK-29997 ? Or, you can revert SPARK-29997 and do this cleanly.
Yes let's do this as a followup of that. I'll modify the title of this PR.

**Comment by sarutak** [comment]
The failed unit tests successfully finished on my laptop.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
I understand that I hit SPARK-30068...

**Comment by HyukjinKwon** [comment]
Seems like:

```
org.apache.spark.sql.hive.thriftserver.ui.ThriftServerPageSuite.thriftserver page should load org.apache.spark.sql.hive.thriftserver.ui.ThriftServerPageSuite.thriftserver session page should load successfully
```

These two tests are being failed globally.

**Comment by sarutak** [comment]
Yeah, exactly. Thanks for the investigation!

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by sarutak** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by sarutak** [comment]
@dongjoon-hyun 
In the current implementation, if we run an empty job, we'll see like a following UI with funny stage id `-1`.
スクリーンショット 2019-12-13 18 13 37

This change fixes it.

File: core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala [hunk 1]
```diff
@@ -756,6 +756,22 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser with Matchers with B
     }
   }
 
+  test("description for empty jobs") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala]: Thank you for adding a test case.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 2]
```diff
@@ -827,7 +827,7 @@ private[spark] object ApiHelper {
   def lastStageNameAndDescription(store: AppStatusStore, job: JobData): (String, String) = {
     // Some jobs have only 0 partitions.
     if (job.stageIds.isEmpty) {
-      ("", job.name)
+      ("", "")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Do we need this change in this PR? UT seems to pass without this.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -695,21 +695,20 @@ private[spark] class DAGScheduler(
...
       listenerBus.post(
-        SparkListenerJobStart(
-          jobId, time, Seq[StageInfo](dummyStageInfo), Utils.cloneProperties(properties)))
+        SparkListenerJobStart(jobId, time, Seq.empty, Utils.cloneProperties(properties)))
       listenerBus.post(
         SparkListenerJobEnd(jobId, time, JobSucceeded))
+
+      if (currentDescription == null) {
+        sc.setJobDescription(null)
+      }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Hi, @sarutak . Shall we avoid changing `sc` status? I made a PR to you.
- https://github.com/sarutak/spark/pull/1

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 2 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Thanks. I've reverted this change.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Thanks! I've merged it.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for reimplementing this, @sarutak .

To reviewers
This followup PR reverts the most part of original PR as a refactoring. Please refer the first one together.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Pull request closed without merging**